### ISSUE TITLE
[GSdx hardware bug] Certain lighting effects pass through solid 3D objects in all of Incognito/Eat Sleep Play's games.

### ISSUE THREAD

**Issue by Aced14** [PCSX2/pcsx2#1276]
**PCSX2 version:**
v1.5.0-dev-504-gf6ee945 (2016-03-28; f6ee94589ffe86ff61d72cf33a425fdf3bc93d4d).

**PCSX2 options:**
Defaults. No options modifications affect this issue.

**Plugins used:**
Defaults (GSdx32-AVX, LilyPad, SPU2-X, cdvdGigaherz, USBnull, FWnull and DEV9ghzdrk). No particular plugins affect this issue.

**Plugin settings:**
GSdx (Direct3D11/OpenGL hardware modes with default settings). No hack modifications affect this issue.

**Description:**
In all of Incognito/Eat Sleep Play's games, certain lighting effects pass through solid objects in GSdx hardware mode. For instance, sun outlines, lens flare effects, car headlights and glowing eyes on opponents.

In my GSdx settings, I've tried toggling between nearly every hardware and OpenGL hack that would be likely to make a difference to no avail. Specifically, I'm referring to CRC hack levels, hardware hacks (half-pixel offset, align sprite, Wild Arms offset, safe accurate blending, preload data frame), hardware depth, accurate date and blending unit accuracy.

None of these issues exist in GSdx software mode. Starting PCSX2 with a hardware renderer, pressing F9 to toggle to software, then F9 again to toggle back to hardware seems to lessen the issue somewhat. I'd recommend against toggling between renderers while testing.

Any chance GSdx's hardware modes could be enhanced to resolve these lighting issues?

**Screenshots:**
* Twisted Metal: Black
  1. [v1.5.0-dev-498-ga99750a (OpenGL software)](https://cloud.githubusercontent.com/assets/13129485/14097255/326f39b0-f53d-11e5-9e1c-fe3792aeaca6.png)
  2. [v1.5.0-dev-498-ga99750a (OpenGL hardware)](https://cloud.githubusercontent.com/assets/13129485/14097254/2fe57c5e-f53d-11e5-9b82-60bb93c5d996.png)
* Twisted Metal: Black Online
  1. [v1.5.0-dev-498-ga99750a (OpenGL software)](https://cloud.githubusercontent.com/assets/13129485/14097644/29b4d52e-f541-11e5-8b73-3adc2cf66671.png)
  2. [v1.5.0-dev-498-ga99750a (OpenGL hardware)](https://cloud.githubusercontent.com/assets/13129485/14097645/2bac8048-f541-11e5-8833-68a88029614f.png)
* War of the Monsters
  1. [v1.5.0-dev-498-ga99750a (OpenGL software)](https://cloud.githubusercontent.com/assets/13129485/14227899/2042d436-f8d5-11e5-98d5-d1538ff30c90.png)
  2. [v1.5.0-dev-498-ga99750a (OpenGL hardware)](https://cloud.githubusercontent.com/assets/13129485/14227900/22e57f7c-f8d5-11e5-88e6-0544be55fc8a.png)
* Downhill Domination
  1. [v1.5.0-dev-498-ga99750a (OpenGL software)](https://cloud.githubusercontent.com/assets/13129485/14097920/a85d1e42-f544-11e5-9474-ea736b0db445.png)
  2. [v1.5.0-dev-498-ga99750a (OpenGL hardware)](https://cloud.githubusercontent.com/assets/13129485/14097921/aa38962e-f544-11e5-9a27-945f0b89fe77.png)
* Twisted Metal: Head-On - Extra Twisted Edition
  1. [v1.5.0-dev-498-ga99750a (OpenGL software)](https://cloud.githubusercontent.com/assets/13129485/14227502/80b5f7d8-f8cd-11e5-9ce8-9c92311ba369.png)
  2. [v1.5.0-dev-498-ga99750a (OpenGL hardware)](https://cloud.githubusercontent.com/assets/13129485/14227501/7e53f2ec-f8cd-11e5-9bc2-3e7715c55375.png)

**GS dump:**
[Twisted Metal: Black in v1.5.0-dev-504-gf6ee945 (OpenGL hardware)](https://Aced14.github.io/pcsx2-gs-dumps/tmb_gs_light-through-objects-bug.7z).

**How to replicate (Twisted Metal: Black):**

1. Launch PCSX2.
2. Open GSdx's plugin settings and select Direct3D11 or OpenGL's hardware renderer.
3. Fast boot the game.
4. Skip the intro FMVs.
5. From the main menu, select "ONE PLAYER - CHALLENGE".
6. Pick any car and select a level called "HIGHWAY LOOP".
7. Select one opponent, then press the O button to go-in-game.
8. Once in game, hold down all 4 shoulder buttons and simultaneously press left, right, down, up on the dpad. This will give you infinite health (to prevent the opponent from quickly killing you).
9. Rotate your car, look for a sun outline, then drive around the level. The sun outline will appear through all objects (such as walls, buildings, bridges and the opponent).
10. Step 9 didn't work as expected. The sun outline is only supposed to appear when its part of the sky is in plain sight. It's not meant to appear through solid objects.

**How to replicate (Twisted Metal: Black Online):**

1. Ensure Ethernet is enabled in the DEV9ghzdrk plugin and that a PS2 network configuration is present on your virtual memory card. This game lacks a built-in network configuration utility.
2. Launch PCSX2.
3. Open GSdx's plugin settings and select Direct3D11 or OpenGL's hardware renderer.
4. Fast boot the game.
5. Skip the intro FMVs and select "LAN" in the main menu.
6. From the LAN lobby, press square, then X to create a game room.
7. In the waiting room, select "GAME SETUP - GAME LEVEL", then pick "HIGHWAY LOOP".
8. Return to the waiting room and select "PLAY GAME".
9. Once in-game, rotate your car, look for a sun outline, then drive around the level. The sun outline will appear through all objects (such as walls, bridges and buildings).
10. Step 9 didn't work as expected. The sun outline is only supposed to appear when its part of the sky is in plain sight. It's not meant to appear through solid objects.

**How to replicate (NTSC-U version of War of the Monsters):**

1. In PCSX2's cheats folder, create a pnach called ``2EDE12D1.pnach``. Place ``patch=1,EE,2015B5C8,extended,00000000 //0C0405E2 - Disable bot AI (nop jal to Ai::updateInputs(void))`` inside the pnach. Without this cheat, it's very difficult to easily replicate the issue at will.
2. Launch PCSX2.
3. Ensure cheats are enabled by going to "System - Enable Cheats".
4. Open GSdx's plugin settings and select Direct3D11 or OpenGL's hardware renderer.
5. Fast boot the game.
6. Skip the intro FMVs.
7. From the main menu, press start, select "1 player - free-for-all - 1 cpu monsters - sudden death", pick Congar (King Kong lookalike) as both player 1 and cpu 1's character, then select the default level ("MIDTOWN PARK").
8. Once in-game, your character will spawn on top of a tall building. Hold down L1 and L2 to lock-onto your opponent. Whilst still locked-on, jump down onto the smaller buildings near where the opponent spawned at.
9. Let go of L1 and L2 to stop locking-on. Your character should continue to face its opponent and the camera will automatically re-align itself. Two small glowing violet-coloured dots (the opponent character's glowing eyes) will be visible through the small buildings and your character.
10. Step 9 didn't work as expected. The opponent's glowing eyes are only supposed to appear if their character is facing you and if its eyes are in plain sight.

**How to replicate (Downhill Domination):**

1. Launch PCSX2.
2. Open GSdx's plugin settings and select Direct3D11 or OpenGL's hardware renderer.
3. Fast boot the game.
4. Skip the intro FMVs.
5. From the main menu, press X, select "SINGLE PLAYER", pick any character, "START NEW GAME", "SINGLE EVENT", then "FREERIDE".
6. At the "INVENTORY" screen, press Start and select the default level ("MT. MiDOULE FR").
7. In-game, once the 3-2-1 countdown timer elapses, turn around and cycle backwards. Your goal is to position your rider directly behind the left side of the mountaintop.
8. Once you're behind the side of the mountain, rotate your cyclist frontwards and ensure the mountain is taking up most of your viewport.
9. You should see a sun outline and a rainbow lens flare effect... through the mountain.
10. Step 9 didn't work as expected. The sun/lens flare effects are only supposed to appear when their part of the sky is in plain sight. They're not meant to appear through solid objects.

**How to replicate (Twisted Metal: Head-On - Extra Twisted Edition):**

1. Launch PCSX2.
2. Open GSdx's plugin settings and select Direct3D11 or OpenGL's hardware renderer.
3. Fast boot the game.
4. Skip the memory card screen and intro FMVs.
5. At the main menu, select "Twisted Metal: Head-On" and skip its intro FMV.
6. Select "SINGLE PLAYER", "ENDURANCE", pick any car and select the "Los Angeles" level.
7. Once in game, hold down all 4 shoulder buttons and simultaneously press left, right, down, up on the dpad. This will give you infinite health (to prevent the opponent from quickly killing you).
8. Rotate your car, look for a sun/lens flare effect, then drive around the level. The sun outline will appear through all objects (such as walls, buildings, bridges and the opponent).
9. Step 8 didn't work as expected. The sun/lens flare effects are only supposed to appear when their part of the sky is in plain sight. They're not meant to appear through solid objects.

**Last known version to work:**
Never worked. Oldest tested version: R5766 (2013-12-05; 9e1717fee542ee1641a6e719bafc403437475d2c).

**PC specifications:**
CPU: Intel Core i7 2600K @ 3.4GHz; GPU: XFX Radeon HD 6970 2GB; OS: Windows 7 SP1.

**Other comments:**
* This issue seems similar to what was described for TimeSplitters in #45, but that involves a different game engine from a different developer... so there's no guarantee it's the same underlying cause.
* FWIW, all offline versions of Twisted Metal: Black and the NTSC-U version of War of the Monsters contain debug symbols in their ELF files.

**Comment by willkuer** [comment]
I guess you can resolve the issue with some large skipdraw value (let's say 100) - however that removes as well most interesting effects in the game and thus is not a satisfying workaround?

**Comment by gregory38** [comment]
I added a log in the code to easily spot the issue. The issue is that EE reads the GS depth buffer. Which mean we need some code to read the GPU depth buffer and send it back to the EE. 

It impacts various game (Nocturne too). GS dump is useless here, so it isn't easy to debug and fix.

**Comment by gregory38** [comment]
Hum could you test the pr with depth texture enabled.

**Comment by Aced14** [comment]
@willkuer That seems to resolve it in certain games, but negatively impacts other unrelated effects (including breaking/hiding HUDs). In Twisted Metal: Black (and presumably Online - didn't test it), I went as high as 200 without seeing any tangible impacts on its sun outline. In every other game I was able to get the effect to partially or fully disappear with high skipdraw values.

@gregory38 Just tried out #1278's [current automated build](https://ci.appveyor.com/api/buildjobs/hhebi62t0d627t05/artifacts/pcsx2-v1.5.0-dev-542-gab94e33-master-793-vs2013-Win32-AppVeyor.7z) to no avail :(. From what I can tell, it didn't have any impact on any of this issue's games.
